docs/guides/spmd_api.md

```diff
@@ -0,0 +1,429 @@
+class DotReluDot(nn.Module):
+  depth: int
+  @nn.compact
+  def __call__(self, x):
+    W1 = self.param(
+        'W1', 
+        nn.with_partitioning(nn.initializers.xavier_normal(), (None, 'model')),
+        (x.shape[-1], self.depth))
+
+    y = jax.nn.relu(jnp.dot(x, W1))
+    # force a local sharding annotation.
+    y = with_sharding_constraint(y, PartitionSpec('data', 'model'))
+
+    W2 = self.param(
+        'W2', 
+        nn.with_partitioning(nn.initializers.xavier_normal(), ('model', None)),
+        (self.depth, x.shape[-1]))
+
+    z = jnp.dot(y, W2)
+    # force a local sharding annotation.
+    z = with_sharding_constraint(z, PartitionSpec('data', None))
+
+    # Return a tuple to conform with the API `nn.scan` as shown in the cell below.
+    return z, None
+
+```
+
+<!-- #region id="1gvf6_a_OeU3" -->
+Now define the `MLP` model as multiple layers of `DotReluDot`.
+
+Note that you can use `nn.scan` and other [lifted transforms](https://flax.readthedocs.io/en/latest/advanced_topics/lift.html) on the module. This will introduce another dimension on the params, and this dimension should be annotated in `metadata_params`. Since the parameters inside `DotReluDot` are already sharded along `model` axis, we don't need to partition multiple layers across `model` dimension here, and therefore denote it as `None`.
+
+To replicate identical layers, you can either use `nn.scan` or a for-loop. `nn.scan` offers faster compilation times, wheras the for-loop is faster on runtime. This guide uses `nn.scan` simply to show that lifted transforms are supported in this API.
+<!-- #endregion -->
+
+```python id="_noFQzrgJ2j0"
+class MLP(nn.Module):
+  num_layers: int
+  depth: int
+  use_scan: bool
+  @nn.compact
+  def __call__(self, x):
+    if self.use_scan:
+      x, _ = nn.scan(DotReluDot, length=self.num_layers, 
+                    variable_axes={"params": 0},
+                    split_rngs={"params": True},
+                    metadata_params={nn.PARTITION_NAME: None}
+                    )(self.depth)(x)
+    else:
+      for i in range(self.num_layers):
+        x, _ = DotReluDot(self.depth)(x)
+    return x
+```
+
+<!-- #region id="pZ6E5btPr5yZ" -->
+# Sharding specification
+
+Now we want to generate the `PartitionSpec` that `pjit` should receive as annotations of input and output data.
+
+The input is relatively simple. For data parallelism, we can shard the batched input `x`s across the `data` axis, by denoting the batch axis as `data`.
+
+<!-- #endregion -->
+
+```python id="63TwLe6vNtID"
+# Input partition rule - data-parallel only
+x_spec = PartitionSpec('data', None)  # dimensions: (batch, length)
+x_spec
+```
+
+<!-- #region id="3d6b15d08980" -->
+However, to generate `PartitionSpec` for the output, we need to use some actual output as reference. One solution is to create a model and evaluate `model.init` abstractly using `jax.eval_shape`, and then use `nn.get_partition_spec` to automatically generate the `PartitionSpec`. 
```

Comment:
"we need to use some actual output as reference" --> don't we need to generate the variable dict structure rather than the model outputs?

Reply:
That works too. I am just trying to show a more complicated example here, in which the pjitted function output is another dataclass instead of the simple variable dict. Added the explanation in doc.